File: Valid Anagram (Easy, #242)/anagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        first_dict = dict()

        for c in s:
            if not c in first_dict:
                first_dict[c] = 0
            first_dict[c] += 1
        
        for c in t:
            if not c in first_dict or first_dict[c] == 0:
                return False
            first_dict[c] -= 1

        return True
    
# This solution has time complexity of O(n) and space complexity of O(1)

File: Valid Anagram (Easy, #242)/anagram.cpp
class Solution {
public:
    bool isAnagram(string s, string t) {
        if (s.size() != t.size())
            return false;
        
        unordered_map<char, int> my_map {};
        for (char c : s) {
            if(my_map.find(c) == my_map.end())
                my_map.insert(make_pair(c, 0));
            my_map[c]++;
        }

        for (char c : t) {
            if (!my_map.contains(c) || my_map.at(c) == 0)
                return false;
            my_map[c]--;
        }

        return true;
    }
};

// This solution has time complexity of O(n) and space complexity of O(1)